aten/src/ATen/AccumulateType.h

```diff
@@ -51,147 +52,231 @@
 
 namespace at {
 
-template <typename T, bool is_cuda>
-struct AccumulateType {};
+template <typename T, c10::DeviceType D>
+struct AccumulateTypeDevice {};
+
+template <>
+struct AccumulateTypeDevice<BFloat16, c10::DeviceType::MPS> {
+  using type = float;
+};
+template <>
+struct AccumulateTypeDevice<Half, c10::DeviceType::MPS> {
+  using type = float;
+};
+template <>
+struct AccumulateTypeDevice<Float8_e5m2, c10::DeviceType::MPS> {
+  using type = float;
+};
+template <>
+struct AccumulateTypeDevice<Float8_e4m3fn, c10::DeviceType::MPS> {
+  using type = float;
+};
+template <>
+struct AccumulateTypeDevice<float, c10::DeviceType::MPS> {
+  using type = float;
+};
+template <>
+struct AccumulateTypeDevice<double, c10::DeviceType::MPS> {
+  using type = float;
+};
+template <>
+struct AccumulateTypeDevice<int8_t, c10::DeviceType::MPS> {
+  using type = int64_t;
+};
+template <>
+struct AccumulateTypeDevice<uint8_t, c10::DeviceType::MPS> {
+  using type = int64_t;
+};
+template <>
+struct AccumulateTypeDevice<char, c10::DeviceType::MPS> {
+  using type = int64_t;
+};
+template <>
+struct AccumulateTypeDevice<int16_t, c10::DeviceType::MPS> {
+  using type = int64_t;
+};
+template <>
+struct AccumulateTypeDevice<int32_t, c10::DeviceType::MPS> {
+  using type = int64_t;
+};
+template <>
+struct AccumulateTypeDevice<int64_t, c10::DeviceType::MPS> {
+  using type = int64_t;
+};
+template <>
+struct AccumulateTypeDevice<bool, c10::DeviceType::MPS> {
+  using type = bool;
+};
+template <>
+struct AccumulateTypeDevice<c10::complex<Half>, c10::DeviceType::MPS> {
+  using type = void;
+};
+template <>
+struct AccumulateTypeDevice<c10::complex<float>, c10::DeviceType::MPS> {
+  using type = void;
+};
+template <>
+struct AccumulateTypeDevice<c10::complex<double>, c10::DeviceType::MPS> {
+  using type = void;
+};
```

Comment:
Not sure what's the best way to assert these currently unsupported dtypes.

Reply:
Good question!
What happens in the void case here?
Otherwise not defining type should give a compilation error if you try to use this version of the template?

Reply:
I was experimenting on whether assigning `void` here could give us some kind of signal (like a dummy value) indicating this type is not supported, but it causes a compilation error indeed. So this is not viable.

Do you have suggestions on this?

Reply:
Isn't compilation error ok? People should never use that!

Reply:
Hmm, the compilation error causes the CI builds to fail: https://hud.pytorch.org/pytorch/pytorch/pull/108248?sha=87fcbd8ab0cbd89126dbe27bc873866ca6315ea5

Reply:
The complex64 support was added recently, and complex128 is not yet supported I think.

https://github.com/pytorch/pytorch/blob/c3945b5f8496f068d56ea1879c10d62ee4ec1801/test/test_mps.py#L10816

Therefore, I set the complex acc types to complex64.